pallets/auctions/src/lib.rs

```diff
@@ -44,16 +42,114 @@ const BID_ADD_BLOCKS: u32 = 10;
+pub struct EnglishAuctionData<T: Config> {
+	pub reserve_price: BalanceOf<T>,
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq)]
+pub struct GeneralAuctionData<AccountId, Balance, BlockNumber, NftClassId, NftTokenId, BoundedVec> {
+	pub name: BoundedVec,
+	pub last_bid: Option<(AccountId, Balance)>,
+	pub next_bid_min: Balance,
+	pub start: BlockNumber,
+	pub end: BlockNumber,
+	pub owner: AccountId,
+	pub token: (NftClassId, NftTokenId),
+}
+
 /// Define type aliases for better readability
-pub type BalanceOf<T> = <<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
-pub type AuctionInfoOf<T> = AuctionInfo<
+pub type GeneralAuctionDataOf<T> = GeneralAuctionData<
 	<T as frame_system::Config>::AccountId,
 	BalanceOf<T>,
 	<T as frame_system::Config>::BlockNumber,
 	<T as pallet_uniques::Config>::ClassId,
 	<T as pallet_uniques::Config>::InstanceId,
+	BoundedVec<u8, <T as pallet_uniques::Config>::StringLimit>,
 >;
 
+pub type BalanceOf<T> = <<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+impl<T: Config> sp_std::fmt::Debug for Auction<T> {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		write!(f, "Auction")
+	}
+}
+pub trait NftAuction<AccountId, AuctionId, BalanceOf, NftAuction> {
+	fn bid(bidder: AccountId, auction_id: AuctionId, value: BalanceOf) -> DispatchResult;
+
+	fn close(auction: &Self) -> DispatchResult;
+}
+
+impl<T: Config> NftAuction<T::AccountId, T::AuctionId, BalanceOf<T>, Auction<T>> for EnglishAuction<T> {
+	fn bid(bidder: <T>::AccountId, auction_id: T::AuctionId, value: BalanceOf<T>) -> DispatchResult {
+		<Auctions<T>>::try_mutate_exists(auction_id, |auction| -> DispatchResult {
+			if let Some(Auction::English(data)) = auction {
+				// Lock / Unlock funds
+				if let Some(ref current_bid) = data.general_data.last_bid {
+					<T as crate::Config>::Currency::remove_lock(AUCTION_LOCK_ID, &current_bid.0);
+				}
+				<T as crate::Config>::Currency::set_lock(AUCTION_LOCK_ID, &bidder, value, WithdrawReasons::all());
+
+				data.general_data.last_bid = Some((bidder, value));
+				// Set next minimal bid
+				let bid_step = Permill::from_percent(BID_STEP_PERC).mul_floor(value);
+				data.general_data.next_bid_min = value.checked_add(&bid_step).ok_or(Error::<T>::BidOverflow)?;
+
+				// Avoid auction sniping
+				let block_number = <frame_system::Pallet<T>>::block_number();
+				let time_left = data
+					.general_data
+					.end
+					.checked_sub(&block_number)
+					.ok_or(Error::<T>::TimeUnderflow)?;
+				if time_left < BID_ADD_BLOCKS.into() {
+					data.general_data.end = block_number + BID_ADD_BLOCKS.into();
+				}
+			}
+
+			Ok(())
+		})
+	}
+
+	fn close(auction: &EnglishAuction<T>) -> DispatchResult {
+		pallet_uniques::Pallet::<T>::thaw(
+			RawOrigin::Signed(auction.general_data.owner.clone()).into(),
+			auction.general_data.token.0,
+			auction.general_data.token.1,
+		)
+		.unwrap_or_default();
```

Comment:
Sure about this ?

Reply:
I think we will remove these unwraps once we migrate to the newest version of pallet_nft with @green-jay

Reply:
no this is not right, as per https://github.com/galacticcouncil/Basilisk-node/pull/203#discussion_r740915565 why not just use "?" ?